pkg/collector/container_hc_collector.go

```diff
@@ -109,23 +116,52 @@ func (c *Collector) updateBPFMetrics() {
 	foundContainer := make(map[string]bool)
 	foundProcess := make(map[uint64]bool)
 	keysToDelete := [][]byte{}
-	var ct ProcessBPFMetrics
-	for it := c.bpfHCMeter.Table.Iter(); it.Next(); {
-		key := it.Key()
-		data := it.Leaf()
+	keys := [][]byte{}
+	values := [][]byte{}
+
+	var (
+		ct ProcessBPFMetrics
+	)
+	// if ebpf map batch get operation is supported we use it
+	if ebpfBatchGet {
+		var batchGetErr error
+		keys, values, batchGetErr = attacher.TableBatchGet(c.bpfHCMeter.Module, c.bpfHCMeter.TableName, uint32(unsafe.Sizeof(ct)), ebpfBatchGetAndDelete /* delete after get */)
+		if batchGetErr != nil {
+			klog.V(1).Infof("failed to get bpf table elemets, err: %v", batchGetErr)
+			ebpfBatchGet = false
+			ebpfBatchGetAndDelete = false
+		}
+	}
+	// if ebpf map batch get operation is not supported we iterate over the map
+	if !ebpfBatchGet {
+		for it := c.bpfHCMeter.Table.Iter(); it.Next(); {
+			key := it.Key()
+			value := it.Leaf()
+			keys = append(keys, key)
+			values = append(values, value)
+		}
+	}
+
+	// iterate over the keys and values
+	for i := 0; i < len(keys); i++ {
+		key := keys[i]
+		data := values[i]
 		err := binary.Read(bytes.NewBuffer(data), utils.DetermineHostByteOrder(), &ct)
 		if err != nil {
 			klog.V(5).Infof("failed to decode received data: %v", err)
 			continue // this only happens if there is a problem in the bpf code
 		}
 
-		// if ebpf map batch deletion operation is supported we add the key to the list otherwise we delete the key
-		if config.EnabledBPFBatchDelete {
-			keysToDelete = append(keysToDelete, key)
-		} else {
-			err = c.bpfHCMeter.Table.Delete(key) // deleting the element to reset the counter values
-			if err != nil {
-				klog.Infof("could not delete bpf table elemets, err: %v", err)
+		// if not deleted during get, delete it now
+		if !ebpfBatchGetAndDelete {
```

Comment:
if this is false, when do we do the delete operation? if will not append to the delete list or delete directly ...

Reply:
@rootfs should this line be `if ebpfBatchGetAndDelete {`

The line 203 is the `!ebpfBatchGetAndDelete` case, right?

Reply:
this line is right, if `ebpfBatchGetAndDelete == true`, all the keys are already deleted in the batch get